src/plugins/data/public/index_patterns/fields/field_list.ts

```diff
@@ -18,65 +18,92 @@
  */
 
 import { findIndex } from 'lodash';
+import { ToastsStart } from 'kibana/public';
 import { IndexPattern } from '../index_patterns';
 import { IFieldType } from '../../../common';
 import { Field, FieldSpec } from './field';
+import { FieldFormatsStart } from '../../field_formats';
 
 type FieldMap = Map<Field['name'], Field>;
 
-export interface IFieldList extends Array<Field> {
+interface FieldListDependencies {
+  fieldFormats: FieldFormatsStart;
+  toastNotifications: ToastsStart;
+}
+
+export interface IIndexPatternFieldList extends Array<Field> {
   getByName(name: Field['name']): Field | undefined;
   getByType(type: Field['type']): Field[];
   add(field: FieldSpec): void;
   remove(field: IFieldType): void;
   update(field: FieldSpec): void;
 }
 
-export class FieldList extends Array<Field> implements IFieldList {
-  private byName: FieldMap = new Map();
-  private groups: Map<Field['type'], FieldMap> = new Map();
-  private indexPattern: IndexPattern;
-  private shortDotsEnable: boolean;
-  private setByName = (field: Field) => this.byName.set(field.name, field);
-  private setByGroup = (field: Field) => {
-    if (typeof this.groups.get(field.type) === 'undefined') {
-      this.groups.set(field.type, new Map());
-    }
-    this.groups.get(field.type)!.set(field.name, field);
-  };
-  private removeByGroup = (field: IFieldType) => this.groups.get(field.type)!.delete(field.name);
+export type CreateIndexPatternFieldList = (
+  indexPattern: IndexPattern,
+  specs?: FieldSpec[],
+  shortDotsEnable?: boolean
+) => IIndexPatternFieldList;
 
-  constructor(indexPattern: IndexPattern, specs: FieldSpec[] = [], shortDotsEnable = false) {
-    super();
-    this.indexPattern = indexPattern;
-    this.shortDotsEnable = shortDotsEnable;
+export const getIndexPatternFieldListCreator: (
+  dependencies: FieldListDependencies
+) => CreateIndexPatternFieldList = ({ fieldFormats, toastNotifications }) => (
```

Comment:
nit: it's a pretty difficult to read. 
what about: 
`export const getIndexPatternFieldListCreator = ({ fieldFormats, toastNotifications }: FieldListDependencies): CreateIndexPatternFieldList => class FieldList`

Reply:
oh... i see, you return instance of created  FieldList  class. Not sure that it's a best solution. 
Could you please open src/plugins/data/public/search/search_service.ts and scroll to 

![image](https://user-images.githubusercontent.com/20072247/80751944-61e1c380-8b33-11ea-89b2-073e89ae0573.png)

Using this approach, you can avoid using any getIndexPatternFieldListCreator =)

Reply:
Discussed this offline with @sulemanof. We decided to leave it as it is